arangod/Replication/ReplicationApplier.cpp

```diff
@@ -32,70 +32,134 @@
- public:
-  ApplyThread(ReplicationApplier* applier, std::unique_ptr<TailingSyncer>&& syncer)
-      : Thread("ReplicationApplier"), _applier(applier), _syncer(std::move(syncer)) {}
+/// @brief common replication applier
+struct ApplierThread : public Thread {
+  public:
+  
+  ApplierThread(ReplicationApplier* applier, Syncer* syncer)
+   : Thread("ReplicationApplier"), _applier(applier), _syncer(syncer) {
+     TRI_ASSERT(_syncer);
+   }
 
-  ~ApplyThread() {
+  ~ApplierThread() {
     {
       MUTEX_LOCKER(locker, _syncerMutex);
-      _syncer.reset();
-    }
-    
-    shutdown(); 
-  }
-
- public:
-  void setAborted(bool value) {
-    MUTEX_LOCKER(locker, _syncerMutex);
-
-    if (_syncer) {
-      _syncer->setAborted(value);
+      delete _syncer;
+      _syncer = nullptr;
     }
+    shutdown();
   }
   
   void run() {
     TRI_ASSERT(_syncer != nullptr);
     TRI_ASSERT(_applier != nullptr);
-
+    
     try {
       setAborted(false);
-      Result res = _syncer->run();
+      Result res = runApplier();
       if (res.fail() && res.isNot(TRI_ERROR_REPLICATION_APPLIER_STOPPED)) {
-        LOG_TOPIC(ERR, Logger::REPLICATION) << "error while running applier for " << _applier->databaseName() << ": "
-          << res.errorMessage();
+        LOG_TOPIC(ERR, Logger::REPLICATION) << "error while running applier thread for "
+        << _applier->databaseName() << ": " << res.errorMessage();
       }
     } catch (std::exception const& ex) {
-      LOG_TOPIC(WARN, Logger::REPLICATION) << "caught exception in ApplyThread for " << _applier->databaseName() << ": " << ex.what();
+      LOG_TOPIC(WARN, Logger::REPLICATION) << "caught exception in ApplierThread for " << _applier->databaseName() << ": " << ex.what();
     } catch (...) {
       LOG_TOPIC(WARN, Logger::REPLICATION) << "caught unknown exception in ApplyThread for " << _applier->databaseName();
     }
-
+    
     {
       MUTEX_LOCKER(locker, _syncerMutex);
       // will make the syncer remove its barrier too
       _syncer->setAborted(false);
-      _syncer.reset();
+      delete _syncer;
+      _syncer = nullptr;
     }
-
+    
     _applier->markThreadStopped();
   }
-
- private:
+  
+  virtual Result runApplier() = 0;
+  
+  void setAborted(bool value) {
+    MUTEX_LOCKER(locker, _syncerMutex);
+    if (_syncer) {
+      _syncer->setAborted(value);
+    }
+  }
+  
+protected:
   ReplicationApplier* _applier;
   Mutex _syncerMutex;
-  std::unique_ptr<TailingSyncer> _syncer;
+  Syncer* _syncer;
 };
 
+/// @brief sync thread class
+struct FullApplierThread final : public ApplierThread {
+  FullApplierThread(ReplicationApplier* applier,
+                    std::unique_ptr<InitialSyncer>&& syncer)
+  : ApplierThread(applier, syncer.get()) {
+    syncer.release();
+  }
+  
+  Result runApplier() override {
+    TRI_ASSERT(_syncer != nullptr);
+    TRI_ASSERT(_applier != nullptr);
+    
+    InitialSyncer* initSync = static_cast<InitialSyncer*>(_syncer);
```

Comment:
I was checking if it is thread-safe here to access `_syncer` without having acquired the `_syncerMutex`.

It seems safe as `_syncer` is only deleted by `run()` (which is the only caller of `runApplier()` so it is the same thread and not a problem) and the destructor of `ApplierThread`, which shouldn't be called while the thread is in `runApplier()`. So it is safe. 
But this is not immediately clear to the reader of the code. Maybe add a comment here so the idea is conveyd to future maintainers?